### ISSUE TITLE
More methods

### ISSUE THREAD

**Issue by qwilbird** [ReactiveX/rxdart#127]
Hi,

awesome work in progress with this library!

Just wondering when you intend to add more methods, like: "distinctBy", "groupBy", "bufferByTime", "delay", "swithMap", "retryWhen" as they would really help!

Thank you and good job again!

**Comment by brianegan** [comment]
Thanks much! Glad ya like it :)

  * `switchMap` is called `flatMapLatest` (the old name). We're planning on changing this to `switchMap` soon!
  * `distinctBy` -- Do `distinct` or `distinctUnique` work for this case? Both methods accept an optional function to determine equality.
  * `groupBy` -- We removed this a while back since Dart was planning on including it in the base Stream class. However, that change was reverted and I'm not sure if it's coming back. If it doesn't come back as part of Dart2, we'll bring it back in!

Need to implement:

  * [ ] `delay`
  * [ ] `retryWhen` 
  * [ ] `bufferByTime`

**Comment by qwilbird** [comment]
Awesome! Any chance they will be added in a future release? :)

What do you recommend for unittesting? With RXJS, we use a lot marble testing

Thank you again!

**Comment by frankpepermans** [comment]
delay was added

**Comment by brianegan** [comment]
`bufferTime` added.

**Comment by brianegan** [comment]
`flatMapLatest` has been renamed to `switchMap` as part of 0.16.5.

**Comment by frankpepermans** [comment]
Loads of new ways to buffer or window streams were added as well

**Comment by escamoteur** [comment]
please add replay()

**Comment by frankpepermans** [comment]
Replay is a bit tricky, since you need to be aware when a new listen happens at the end of the downstream.
That's only possible if the `Stream` which is being transformed, is a broadcast `Stream`.

i.e.
```
stream
    .replay()
    .asBroadcastStream(); // will not work with multiple listeners, as the transform happens only once
```

```
stream
    .asBroadcastStream()
    .replay(); // works as expected
```

The safe approach is to always throw when the `Stream` provided to replay is not a broadcast `Stream`

**Comment by brianegan** [comment]
@frankpepermans Yah, `replay` is actually a pretty big topic in itself. I was originally thinking we could essentially convert the Stream into a `ReplaySubject` in the `replay` transformer, but Rx has some interesting conventions around replay and how it should work (it returns a `ConnectableObservable`, for example).

**Comment by brianegan** [comment]
Hrm, that's interesting, I'll have to take a look. I assumed we would only want the transformer called once. The transformer would be responsible for listening to the original stream (just once) and then piping all events to the `ReplaySubject`.

Once it's a ReplaySubject stream, it should be broadcast by default and everyone who subscribes should be all good?

**Comment by long1eu** [comment]
What about this implementation of retryWhen https://gist.github.com/long1eu/9c9e257f111e0c8d3ba44cbf840a8767

**Comment by brianegan** [comment]
Thanks @long1eu! I think we'll need to make this a `Stream` class rather than a `Transformer`, since `retry` / `retryWhen` should re-subscribe to the original Stream rather than switching to a new Stream when the original Stream fails. Since this won't work in Dart as Streams are single-subscription, the StreamFactory needs to be the "Source of Truth" -- it should create the original Stream that could fail, and should re-create that same Stream using the Factory when it does.

**Comment by long1eu** [comment]
I've tried to build it based on your suggestions on top of the retry count implementation.

https://gist.github.com/long1eu/db508dcf2c1a28ec226a939ca5fe28e0

**Comment by qwilbird** [comment]
@long1eu That looks nice - looking forward being able to use retryWhen!
@brianegan, what do you think of this implementation?

**Comment by brianegan** [comment]
Oh dang, sorry I missed this one. Overall -- Looking much better! Thanks for taking the time to write that up :)

I think it would make sense to extract the `ErrorAndStackTrace` class out into it's own file in the `lib/src/utils` and share it between `retry` and `retryWhen`.

It'd also be great to have Docs + Tests for this code so we can let folks know how they can use this constructor and so we verify it works correctly now and into the future. You can check out the retry tests for an example of how to write these types of tests, or if ya need help please let me know.

Once all that is done, it'd be great to submit this work as a PR! Thanks again for all the work :)

**Comment by brianegan** [comment]
Hey all -- just published version 0.18.1 that includes `retryWhen` -- Thanks so much to @long1eu for contributing 🔥 🎉 ❤️ 

I'm going to go ahead and close this issue out now since we've covered the original 3 operators. The `retry` work has been split out into it's own issue for tracking since it's a larger chunk of work.

**Comment by qwilbird** [comment]
awesome work guys! Thank you so much!
